pkg/credentialwatcher/secretwatcher.go

```diff
@@ -0,0 +1,129 @@
+	return &secretWatcher{
+		watchInterval: watchInterval,
+		client:        client,
+	}
+}
+
+// SecretWatcher will trigger CredentialsRotator every `scanInternal` and only stop if the operator is killed or a
+// message is sent on the stopCh
+func (s *secretWatcher) Start(log logr.Logger, stopCh <-chan struct{}) {
+	log.Info("Starting the secretWatcher")
+	for {
+		select {
+		case <-time.After(s.watchInterval):
+			log.Info("secretWatcher: scanning secrets")
+			err := s.ScanSecrets(log)
+			if err != nil {
+				log.Error(err, "secretWatcher not started, credentials wont be rotated")
+			}
+		case <-stopCh:
+			log.Info("Stopping the secretWatcher")
+		}
+	}
+}
+
+// CredentialsRotator will list all secrets with the `STSCredentialsSuffix` and mark the account CR `status.rotateCredentials` true
+// if the credentials CreationTimeStamp is within `STSCredentialsRefreshThreshold` of `STSCredentialsDuration`
+func (s *secretWatcher) ScanSecrets(log logr.Logger) error {
+	// List STS secrets and check their expiry
+	secretList := &corev1.SecretList{}
+
+	listOps := &client.ListOptions{Namespace: awsv1alpha1.AccountCrNamespace}
+	if err := s.client.List(context.TODO(), listOps, secretList); err != nil {
+		log.Error(err, fmt.Sprintf("Unable to list secrets in namespace %s", awsv1alpha1.AccountCrNamespace))
+		return err
+	}
+
+	for _, secret := range secretList.Items {
+
+		accountName := strings.TrimSuffix(secret.ObjectMeta.Name, STSCredentialsSuffix)
+
+		if strings.HasSuffix(secret.ObjectMeta.Name, STSCredentialsSuffix) {
+			unixTime := time.Unix(secret.ObjectMeta.CreationTimestamp.Unix(), 0)
+			timeSinceCreation := int(time.Since(unixTime).Seconds())
+
+			if STSCredentialsDuration-timeSinceCreation < STSCredentialsThreshold {
+
+				accountInstance, err := s.GetAccount(accountName)
+				if err != nil {
+					log.Error(err, "Unable to rerieve account CR %s", accountName)
+				}
+
+				if accountInstance.Status.RotateCredentials != true {
+
+					log.Info(fmt.Sprintf("AWS credentials secret %s was created %s ago requeing to be refreshed", secret.ObjectMeta.Name, time.Since(unixTime)))
+
+					accountInstance.Status.RotateCredentials = true
+
+					err = s.UpdateAccount(accountInstance)
+					if err != nil {
+						log.Error(err, fmt.Sprintf("Error updating account %s", accountName))
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// GetAccount retrieve account CR
+func (s *secretWatcher) GetAccount(accountName string) (*awsv1alpha1.Account, error) {
+	accountInstance := &awsv1alpha1.Account{}
+	accountNamespacedName := types.NamespacedName{Name: accountName, Namespace: awsv1alpha1.AccountCrNamespace}
+
+	err := s.client.Get(context.TODO(), accountNamespacedName, accountInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return accountInstance, nil
+}
+
+// UpdateAccount updates account CR
+func (s *secretWatcher) UpdateAccount(account *awsv1alpha1.Account) error {
```

Comment:
This function or one similar already exists in account_controller.go . We should unify these usages.

Reply:
Its a different package though, I was trying to keep them separate.

Reply:
I realized this after the fact :)